grpc-api/server/docker/build.gradle

```diff
@@ -66,10 +66,18 @@ docker {
   dockerCreateDockerfile {
     //volume("/app/resources")
 
+    runCommand('''set -eux;\\
+                  GRPC_HEALTH_PROBE_VERSION=v0.3.1;\\
+                  curl -L -o /bin/grpc_health_probe https://github.com/grpc-ecosystem/grpc-health-probe/releases/download/${GRPC_HEALTH_PROBE_VERSION}/grpc_health_probe-linux-amd64; \\
+                  chmod +x /bin/grpc_health_probe
+                  ''')
+
     copyFile('licenses/', '/')
 
     volume("/data")
     volume("/cache")
+
+    instruction('HEALTHCHECK --interval=3s --timeout=10s CMD /bin/grpc_health_probe -addr=localhost:8080 -connect-timeout=10s')
```

Comment:
Will fix exit codes, good find - I had an earlier revision that normalized those, but I misunderstood how healthcheck worked and didnt notice an issue afterwards.

Timeouts: my quick testing found that startup times are just under 3 seconds locally, and `interval` is used as the delay until the first attempt is made. I set both docker's timeout grpc-health-probe's connect-timeout to 10s as deliberately excessive, so that after three seconds, we will wait at most 10 seconds before pausing for three seconds and trying again. 

My very unscientific test of --timeout vs -connect-timeout is that after "connect" is successful, it takes at most 10ms (and usually 5ms), to get a healthy reply, so a race hardly matters at all.

Reducing retries to 1 runs the risk of it taking longer than 11 seconds (interval=3 + connect-timeout=8) on some machine to start the server the first time, and websocket/envoy will fail to start at all during docker-compose up. I have a decently fast machine, so we need some extra buffer.

The max possible startup time should be less than `retries * (min(timeout, connect-timeout) + interval)`. I'm not sure where you want to add the extra buffer there, but I think we should allow for slow startups. It appears that app mode _must_ finish starting all of its scripts (serially) before gRPC can start at all, so take my local 3s, call it 10s to support arbitrary hardware/configurations, and add some extra factor for the slowest script we think is reasonable...

Another option would be to let grpc start sooner, but signal that it isn't yet ready for work (error code=4), and accept that as "healthy", and require clients to query it also? 

--

I think the motivation here needs to be "enough information to start downstream workers/containers/clients" and not "instant feedback on status of the server". I initially tuned these to "get to HEALTHY as quickly as possible", since being slow to report UNHEALTHY won't stop downstream clients/containers, it will just provide more evidence of where the problem is. If anything, from revisiting the above params, I would like to increase the retries or the timeouts, so that we get a fast startup and slow failure, and support slow startup app scripts (consuming days worth of kafka streams for example).

Reply:
As per our slack chat, I've tried both reducing HEALTHCHECK timeout to 1s with grpc_health_probe at 10s, and vice versa, so that both the grpc_health_probe times out and fails, and the HEALTHCHECK times out waiting for grpc_health_probe, and in both cases the process stays in "starting" health state, so it appears to be a difference without a distinction?

From the docker documentation on HEALTHCHECK, I'm appending a `|| exit 1` to the end of the command, so that it correctly normalizes all non-zero exit codes to 1.

Reply:
The last 5 health check statuses are recorded by docker and available via docker inspect, and they do allow visibility into where the failure took place, but the actual status and the downstream processes are not impacted at all by this difference.